### ISSUE TITLE
[BUG:] Nullability error in the generated spec UserStatusStatus.clearAt

### ISSUE THREAD

**Issue by Leptopoda** [provokateurin/nextcloud-neon#193]
Yesterday I hit some corner cases within the spec.

The test effectively expects that the `UserStatusStatus.clearAt` is null although it isn't marked as nullable in the spec but still required.

https://github.com/provokateurin/nextcloud-neon/blob/e690ed2456ba508b5789a7a9f1a20b93aa221c17/packages/nextcloud/test/user_status.dart#L46-L48

https://github.com/provokateurin/nextcloud-neon/blob/e690ed2456ba508b5789a7a9f1a20b93aa221c17/packages/nextcloud/lib/src/nextcloud.openapi.json#L1774-L1835

The problem is that `UserStatusClearAt` does not have any requied fields either thus a valid serialized value would be `null` but `build_value` does not like the serialized object being null (and again the `UserStatusStatus.clearAt` is not marked as nullable).

https://github.com/google/built_value.dart/blob/5c71eee4a8890de2bbae85b798be1b9e8d171ce4/built_value/lib/serializer.dart#L336-L337
 
The screenshot shows the output of my vscode debugger.
Ideally the `UserStatusStatus.clearAt` should be marked as nullable. 

![Screenshot_20230214_124421](https://user-images.githubusercontent.com/25266387/218730440-ee2b758b-f960-4407-aa29-625488ab8249.png)

Am I right with my suspicion that the spec is wrong?

**Comment by provokateurin** [comment]
You are right, `clearAt` should be nullable. The current implementation should definitely work because `anyOf` will just not match to anything if the value is null, but I see that the `clearAt` property itself should be nullable to match what is actually returned. You can go ahead and mark `clearAt` as not required :)

**Comment by Leptopoda** [comment]
> The current implementation should definitely work because anyOf will just not match to anything if the value is null

According to the OpenApi documentation anyOf must resolf to at least one value:
> oneOf matches exactly one subschema, and anyOf can match one or more subschemas.

https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/#anyof

So we should add an assertion like the others
```dart
if (schema.anyOf != null) ...[
  "assert([${fields.values.join(',')}].where((final x) => x != null).length > 0, 'Need anyOf for \$data');",
  ],
```

I just added this assertion and it breaks quite a bit more than the `clearAt` :confused: 


I don't think it's necessary to backport (make a separate PR) but I'll fix it on my branch

**Comment by provokateurin** [comment]
I think the correct thing is just making clearAt nullable and use oneOf, right?

**Comment by Leptopoda** [comment]
Probably if we are sure that it really is oneOf and not anyOf (i.e the server never returns both at the same time)

I'm also facing issues in implementing nullable 'ofs' the way I did yesterday, so I might need to scrap this and do it in a different way.

So I need to investigate further and have a look (also at the other fields currently having issues with my assertion.
To be clear the my current code works fine without the assertion but as mentioned it wouldn't really be OpenApi compliant :upside_down_face:

**Comment by provokateurin** [comment]
anyOf is definitely not possible because it can only return one of them.

**Comment by provokateurin** [comment]
Btw I also realized that there are some mistakes in the user_status and provisioning_api spec when I extracted those automatically with my tool. It's not really a problem though because the currents specs work just fine right now and everything is still alpha level anyway.

**Comment by Leptopoda** [comment]
Iirc those where the other two problem spots I found :)

Could you please tell me what the expected behavior would be? I hope I can find time soon to look into it.

**Comment by provokateurin** [comment]
See https://github.com/provokateurin/nextcloud-neon/tree/feature/openapi-automatic. The output spec is formatted and ordered differently, so direct comparison isn't that easy, but you can just look at the components that are defined. Be aware that the specs are not final, I know some things I want to change (e.g. enums for the status again), but they are definitely more correct than the old ones.

**Comment by Leptopoda** [comment]
Ok I rebased onto the openapi-automatic branch and everything was more or less working instantly :)

I just noticed that all current uses of `anyOf` in the spec are a bit off as they only contain one member thus the usage of anyOf is unnecessary. Also It looks like in all three cases it should be `oneOf[List<Type>, Type`.

I just patched the big nextcloud.openapi.json so nothing final and one test is still failing but that's something wired on my side. I'll clean my commits up a bit and you can hopefully expect an initial PR today.

Closing this as the reported issue is solved on the new branch at least.

**Comment by provokateurin** [comment]
^ The problem is that the php returns `[]` in some cases which is an array in json. This really sucks because we expect an object.

**Comment by Leptopoda** [comment]
> This really sucks because we expect an object.

Then what's the problem with just marking it as an Object?  That's what I did in the built_value PR.

shall we reopen this issue?

**Comment by provokateurin** [comment]
Mh yeah, the problem is the result is either an object or an array, but the array value is useless because it is empty anyway

**Comment by Leptopoda** [comment]
That's bad :(
So if I understand correctly the property is either an Object or an empty array and never null but the Object itself isn't an array?
So the empty array represents something like a null value?

I think we could do multiple things:
- fix it on the php side :) I'm not sure why someone would indicate nullability of an Object with an empty List while the Object itself isn't a list. Could take time though changing the server code and now that NC26 is hitting RC state we would at least have to wait another 4 months for 27.
- propperly document it as `oneOf[array[itemCount: 0], Object]` and leave as is (bad for consumers of the dart lib though)
- we could try to detect this state in dynamite and just make a field that's a nullable Object. We could then make our own serializer for that which maps the empty array to null. A problem here it that built_value will decide the nullability of an object before trying to deserialize it. This meas that if the json contains a key it'll try to deserialize it and won't accept null to be returned from the deserialization. We might get arround this by implementing our own serializer type or moving the differentiation one level up so it wouldn't even try to deserialize it.
- just mark it as an object and let the consumer do all the stuff on it that's needed (not really a fan of it)
- finally we could have an ugly object with the oneOf stuff and a helper function like a getter that'll either return the object or null if it is an array (probably the easiest and cleanest to do)

**Comment by provokateurin** [comment]
I think fixing it on the php side is really hard, but not impossible. We would need to be very careful with keeping compatibility and that will be really hard. I think such changes might not be accepted :/

**Comment by provokateurin** [comment]
It's probably best to raise an issue in the server repo and say why the current behaviour is bad and explain that it has to be fixed. We definitely can't fix everything at once, but just fixing it in parallel with making the code ready for automatic specs is fine.
What do you think?

**Comment by Leptopoda** [comment]
Obviously I'd also prefer it fixed in the server itself as the current situation seems quite bad.

> We would need to be very careful with keeping compatibility and that will be really hard.

Why would we need to keep compatibility? Aren't breaking changes allowed? As long as it isn't used in federation I don't see any problems with breaking stuff as long as it is communicated early enough.

**Comment by provokateurin** [comment]
Yeah it's fine, but we need to make sure it's fixed in the frontend too (although I'm not sure if the result isn't ignored quite often anyway)

**Comment by provokateurin** [comment]
Fixing it on the server also means that 26 or higher is the minimum required version. I don't think something like this is applicable for backporting since it's not a critical bug fix

**Comment by Leptopoda** [comment]
Judging from the nc-android client that still supports nextcloud 16 theay surely have some sort of backwards compatibility. I can't imagine there haven't been any changes in the last 10 versions.
Although nextcloud is usually great at keeping backwards compatibility.

I think it is best if we just raise an issue and see how far we can get :)

**Comment by provokateurin** [comment]
I'll first discuss this with some colleagues, since it might just be part of my work if deemed necessary. Then we don't need to raise an issue and I can just go about it and fix it everywhere (since such a big change might be harder to make it through review).

**Comment by provokateurin** [comment]
Ok, so we can't fix it in the server because exisiting code relies on it not being null.

I'd like to avoid making dynamite specific to this project.
I'll your idea of going with `oneOf[array[itemCount: 0], Object]` since this is the correct specification. In dynamite we can the filter out the arrays that have `itemCount: 0` and convert `oneOf`s with a single option into the option but with nullable. This seems quite neutral to me and the best solution without changing server code.

**Comment by Leptopoda** [comment]
nevertheless it might be worth raising internally to at least raise awareness of this issue for the future

I think this would fit to other 'technical debt' issues
https://github.com/nextcloud/server/issues?page=1&q=is%3Aopen+is%3Aissue+label%3A%22technical+debt%22

**Comment by Leptopoda** [comment]
I've made a simple POC of how this could look.
https://github.com/Leptopoda/OpenApi-Dart-Nextcloud-OneOf

just what I could come up with in 10 minutes or so :)

This is at least my idea. Obviously We'd only need to write a custom serializer for the classes needing it and not for every.

**Comment by provokateurin** [comment]
Looks interesting!

**Comment by Leptopoda** [comment]
As you can see I just used the generated serializer and added a simple condition.

To make it even simpler we might be able to just extend the generated serializer adding our condition and call `super.deserialize` when we don't hit our special case. 

Thus we wouldn't need to write our own serializer.

**Comment by provokateurin** [comment]
Yeah I'd like to add as little special sauce as possible.

**Comment by Leptopoda** [comment]
> Ok, so we can't fix it in the server because exisiting code relies on it not being null.

new Idea:
we can't make it null but we can return an empty string `''`. And that already is a valid response judging from

https://github.com/nextcloud/server/blob/61679b4fdb61f7443b00863b575846ccf94f1e08/core/Controller/AutoCompleteController.php#L106

Also the test only uses empty strings :woman_shrugging: 
https://github.com/nextcloud/server/blob/61679b4fdb61f7443b00863b575846ccf94f1e08/tests/Core/Controller/AutoCompleteControllerTest.php

My php knowledge is quite bad so I might be wrong

**Comment by provokateurin** [comment]
I think we can't do anything about it, because clients expect a certain format. No matter what we'd change, we would need to adapt everything to it.
